src/util/PageRange.cpp

```diff
@@ -1,76 +1,110 @@
-
-auto PageRangeEntry::getFirst() const -> int { return this->first; }
-
-auto PageRange::isSeparator(char c) -> bool { return (c == ',' || c == ';' || c == ':'); }
-
-auto PageRange::parse(const char* str, int pageCount) -> PageRangeVector {
-    PageRangeVector data;
-
-    if (*str == 0) {
-        return data;
+#include <charconv>
+#include <cstddef>
+#include <regex>
+#include <stdexcept>
+#include <string>
+#include <utility>
+
+
+/**
+ * @brief Parse a string of page ranges.
+ *
+ * This function parses a string of page ranges into a vector of pairs of page
+ * numbers. A page range is of the form n, n-, -m, n-m where n, m are positive
+ * integers. The input - is also accepted. Page ranges are separated by `,`, `;`
+ * and `:`. Whitespace is ignored. The parameter 'pageCount' is the largest page
+ * number that may be refered to. The function silently corrects page numbers
+ * equal to 0 or larger than pageCount by truncating to the nearest
+ * acceptable integer.
+ *
+ * Example input and output:
+ *
+ *     parse("1, 2-, -3, 4-5, -, 0-42", 10)
+ *     ===>
+ *     {0, 0}, {1, 9}, {0, 2}, {3, 4}, {0, 9}, {0, 9}
+ *
+ * Note that the page numbers are parsed in the format 1 - pageCount
+ * (permissive) and the return values are in the range 0 - (pageCount-1).
+ *
+ * @param s The string containing page ranges to parse.
+ * @param pageCount The largest page that is possible to refer to.
+ * @exception std::logic_error If `pageCount == 0`.
+ * @exception std::invalid_argument If the input doesn't match any acceptable
+ * page ranges.
+ * @return A vector containing the page ranges.
+ */
+std::vector<PageRangeEntry> PageRange::parse(const std::string& s, size_t pageCount) {
+    // break the string into comma (or ;:) separated tokens.
+    const std::regex separators("[,;:]");
+    auto begin = std::sregex_token_iterator(s.cbegin(), s.cend(), separators, -1);
+    auto end = std::sregex_token_iterator();
+
+    std::vector<PageRangeEntry> entries;
+    if (pageCount == 0) {
+        throw std::logic_error("PageRange::parse(): pageCount is zero.");
     }
 
-    int start = 0, end = 0;
-    char* next = nullptr;
-    const char* p = str;
-    while (*p) {
-        while (isspace(*p)) {
-            p++;
-        }
-
-        if (*p == '-') {
-            // a half-open range like -2
-            start = 1;
+    /* The following input cases are considered:
+     *
+     *   1) n,   parsed by singlePage,
+     *   2) n-,  parsed by rightOpenRange,
+     *   3) -m,  parsed by leftOpenRange,
+     *   4) n-m, parsed by pageRange,
+     *   5) -,   parsed by bothOpenRange,
+     *   6) For everything else, an exception is thrown.
+     */
+    const std::regex singlePage("\\s*(\\d+)\\s*");
+    const std::regex rightOpenRange("\\s*(\\d+)\\s*-\\s*");
+    const std::regex leftOpenRange("\\s*-\\s*(\\d+)\\s*");
+    const std::regex pageRange("\\s*(\\d+)\\s*-\\s*(\\d+)\\s*");
+    const std::regex bothOpenRange("\\s*-\\s*");
```

Comment:
I think this is unnecessarily complicated. Why not a regex with three capture groups?
```
\\s*(\\d+)?\\s*(-)?\\s*(\\d+)?\\s*
```

Reply:
I would argue that the code documents itself. Originally it was like you suggested, I opted for the other option because the resulting code further below was cleaner.